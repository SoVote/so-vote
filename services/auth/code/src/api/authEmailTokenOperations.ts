import crypto from "crypto";
import { ssm } from '@rainbow-husky/common-backend'

// const key = Buffer.from('8039ace2aea257276dc0fedcfd720a221f5d86603f0a951cbe2d87feca8ece8f' as string, 'hex');
const iv = Buffer.from('fa646872cfa756d22966c1934f3a8ae3f9fdae6a', 'hex');
const authTagLength = 16

const authSecretName = `rh-shared-${process.env.BRANCH === 'main' ? 'prod' : 'dev'}-auth-secret`

interface TokenBody {
  email: string,
  issued: string
}

export const generateAuthEmailToken = async (email: string) => {
  const key = await ssm.getCachedSsmParameter(authSecretName)
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv, { authTagLength })
  const body: TokenBody = { email, issued: new Date().toISOString() }
  return Buffer.concat([ cipher.update(JSON.stringify(body)), cipher.final(), cipher.getAuthTag()]).toString('hex')
}

export const parseAuthEmailToken = async (token: string): Promise<TokenBody> => {
  const key = await ssm.getCachedSsmParameter(authSecretName)
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv, { authTagLength });
  const tokenBuffer = Buffer.from(token, 'hex')
  const bodyCutoffPosition = tokenBuffer.length - authTagLength
  const encryptedBody = tokenBuffer.subarray(0, bodyCutoffPosition)
  const authTag = tokenBuffer.subarray(bodyCutoffPosition, tokenBuffer.length)
  decipher.setAuthTag(authTag);
  return JSON.parse(Buffer.concat([ decipher.update(encryptedBody), decipher.final()]).toString('utf-8'))
}