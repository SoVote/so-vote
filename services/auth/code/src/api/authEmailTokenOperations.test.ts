import { generateAuthEmailToken, parseAuthEmailToken } from "./authEmailTokenOperations";

beforeEach(() => {
  jest.resetAllMocks()
  jest.useFakeTimers({ now: new Date('2024-01-01') })
})

afterAll(() => {
  jest.useRealTimers()
})

describe('authEmailTokenOperations', () => {
  it('should encrypt and decrypt valid tokens correctly', () => {
    const encryptedData = generateAuthEmailToken('some@email.address')
    const decryptedDataObject = parseAuthEmailToken(encryptedData)
    expect(decryptedDataObject).toMatchSnapshot()
  })
  it('should throw an error if the token is invalid', () => {
    const mangledToken = '13d01a9ecb107ed564ec1423e499d26c2781d413f92949c7973f09be7924d82ae8aaa1a34bd439b08174f2b7affcbbf452f5d19c3a6f660ffbc080ae2c938292c85eb9652798761c6e7afbf8cc75e75686db'
    expect(() => parseAuthEmailToken(mangledToken)).toThrowErrorMatchingSnapshot()
  })
})