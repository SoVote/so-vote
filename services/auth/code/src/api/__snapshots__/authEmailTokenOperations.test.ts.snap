// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authEmailTokenOperations should throw an error if the token is invalid 1`] = `"Unsupported state or unable to authenticate data"`;
